### PR TITLE
[Android] mem ptr to byte-buffer

### PR DESCRIPTION
**Pull request by jaeyun-jung**
In native code, we can get the mem ptr of byte-buffer.
Remove memcopy to convert tensors object and data struct in native.

Signed-off-by: Jaeyun <jy1210.jung@samsung.com>

### CODE REVIEW

**Comment by jaeyun-jung** [comment]
@anyj0527 please check the latency with this PR.

**Comment by myungjoo** [comment]
> @anyj0527 please check the latency with this PR.

If this makes meaningful differences in latency, could you please show the differences here?

**Comment by anyj0527** [comment]
@myungjoo @jaeyun-jung 

With NNFW sub-plugin.
Used model: ssd_mobilenet_v2_coco
input: (float32)1x300x300x3 (1,080,000 bytes)
output: (float32)1x1917x1x4, (float32)1x1917x91 (total 728,460 bytes)


before this patch:

2020-08-04 19:30:40.038 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: count: 100
2020-08-04 19:30:40.039 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: prepare: 1817
2020-08-04 19:30:40.039 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: invoke: 653053
2020-08-04 19:30:40.039 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: result: 1991
2020-08-04 19:30:40.039 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: total: 656861
2020-08-04 19:30:40.039 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-java-api: latency of `TensorsData SingleShot.invoke(TensorsData in)`: 657616

ml_single_invoke: 653053
native_single_invoke: 656861, diff: 3808
SingleShot.invoke: 657616, diff: 755
**total diff: 4563**

==================================================================================
after this patch:

2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: count: 100
2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: prepare: 157
2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: invoke: 656682
2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: result: 496
2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: total: 657336
2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-java-api: latency of `TensorsData SingleShot.invoke(TensorsData in)`: 658090

ml_single_invoke: 656682
native_single_invoke: 657336, diff: 654
SingleShot.invoke: 658090, diff: 754
**total diff: 1408**

==================================================================================

The overhead of `nns_native_single_invoke` reduces from 4563 us to 1408 us. Great!

File: api/android/api/src/main/jni/nnstreamer-native-api.c [hunk 1]
```diff
@@ -463,16 +461,15 @@ nns_parse_tensors_data (pipeline_info_s * pipe_info, JNIEnv * env,
...
-      if (data->tensors[i].tensor == NULL) {
-        nns_loge ("Failed to allocate memory %zd, index %d.", data_size, i);
-        (*env)->DeleteLocalRef (env, tensor);
-        failed = TRUE;
-        goto done;
+      if (clone) {
+        if (data->tensors[i].tensor == NULL)
+          data->tensors[i].tensor = g_malloc (data_size);
+
+        memcpy (data->tensors[i].tensor, data_ptr, data_size);
```
Comment (anyj0527) [inline api/android/api/src/main/jni/nnstreamer-native-api.c:468]: Need to handle memory error case here?

Comment (jaeyun-jung) [reply api/android/api/src/main/jni/nnstreamer-native-api.c:468]: low mem case, g_malloc will break if failed to allocate mem size :)

Comment (anyj0527) [reply api/android/api/src/main/jni/nnstreamer-native-api.c:468]: Oh right!

**Comment by myungjoo** [comment]
> @myungjoo @jaeyun-jung
> 
> Tested on S20 (QC) with NNFW sub-plugin
> Used model: ssd_mobilenet_v2_coco
> input: (float32)1x300x300x3 (1,080,000 bytes)
> output: (float32)1x1917x1x4, (float32)1x1917x91 (total 728,460 bytes)
> 
> invoke 100 times and print average time in microsecond
> prepare : singleshot-invoke called, prepare input data, convert tensors (object -> data handle). Mainly for `nns_parse_tensors_data`
> invoke : ml_single_invoke()
> result : prepare invoke result, convert tensors (data handle -> object). Mainly for `nns_convert_tensors_data`
> native total: nns_native_single_invoke()
> java api total: TensorsData SingleShot.invoke(TensorsData in)
> 
> before this patch:
> 
> 2020-08-04 19:30:40.038 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: count: 100
> 2020-08-04 19:30:40.039 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: prepare: 1817
> 2020-08-04 19:30:40.039 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: invoke: 653053
> 2020-08-04 19:30:40.039 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: result: 1991
> 2020-08-04 19:30:40.039 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: total: 656861
> 2020-08-04 19:30:40.039 21966-22016/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-java-api: latency of `TensorsData SingleShot.invoke(TensorsData in)`: 657616
> 
> ml_single_invoke: 653053
> native_single_invoke: 656861, diff: 3808
> SingleShot.invoke: 657616, diff: 755
> **total diff: 4563**
> 
> ==================================================================================
> after this patch:
> 
> 2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: count: 100
> 2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: prepare: 157
> 2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: invoke: 656682
> 2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: result: 496
> 2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: total: 657336
> 2020-08-10 10:00:37.055 31178-31252/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-java-api: latency of `TensorsData SingleShot.invoke(TensorsData in)`: 658090
> 
> ml_single_invoke: 656682
> native_single_invoke: 657336, diff: 654
> SingleShot.invoke: 658090, diff: 754
> **total diff: 1408**
> 
> ==================================================================================
> 
> The overhead of `nns_native_single_invoke` reduces from 3808 us to 654 us.
> The overhead of `SingleShot.invoke` keeps similar results (755 us, 754 us).
> Total: 4563 us -> 1408 us. Great!

Note: this latency value is "per 100 executions", not "per execution".
Thus, per each invoke, the latency has become 45.63us --> 14.08us (Java) / 38.08 us --> 6.54 us (C/C++)

**Comment by myungjoo** [comment]
@jaeyun-jung We need to update the performance report for Android releases, too :)

**Comment by anyj0527** [comment]
@myungjoo @jaeyun-jung 

**mobilenet with TF-lite sub-plugin**
input: (uint8)1x224x224x3 = 150,528 bytes
output: (uint8)1x1001 = 1,001 bytes

2020-08-13 10:58:28.624 22157-22217/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: count: 100
2020-08-13 10:58:28.625 22157-22217/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: prepare: 42
2020-08-13 10:58:28.625 22157-22217/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: invoke: 47870
2020-08-13 10:58:28.625 22157-22217/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: result: 91
2020-08-13 10:58:28.625 22157-22217/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: total: 48004
2020-08-13 10:58:28.625 22157-22217/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-java-api: latency of `TensorsData SingleShot.invoke(TensorsData in)`: 48187

ml_single_invoke: 47870
native_single_invoke: 48004, diff: 134
SingleShot.invoke: 48187, diff: 183
total diff: 317

======================================================
previous result from https://github.com/nnstreamer/nnstreamer/issues/2165#issuecomment-647444476
2020-06-22 19:37:26.541 15702-15746/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: count: 100
2020-06-22 19:37:26.541 15702-15746/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: prepare: 219
2020-06-22 19:37:26.541 15702-15746/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: invoke: 47940
2020-06-22 19:37:26.541 15702-15746/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: result: 426
2020-06-22 19:37:26.541 15702-15746/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-native: total: 48587
2020-06-22 19:37:26.541 15702-15746/org.nnsuite.nnstreamer.test W/nnstreamer-benchmark-java-api: latency of `TensorsData SingleShot.invoke(TensorsData in)`: 48940

ml_single_invoke: 47940
native_single_invoke: 48587, diff: 647
SingleShot.invoke: 48940, diff: 353
total diff: 1000

Overall overhead: 1000 us -> 317 us

**Pull request merged**